WordPress/gutenberg
# Constrained tabbing: fix unstable behavior in firefox

Fix #42652, #45903
Follow up on #34836

## What?
This PR fixes a problem with "constrained tabbing" sometimes losing focus in Firefox.

"Constrained tabbing" means that the tab moving is constrained to the content of the wrapper element such as modal component.

As reported in #42652, I confirmed that in Firefox, the focus is sometimes lost from the modal when the following operation is performed:

- When `tab` key is pressed on the **last** focusable element (the first element should be focused)
- When `tab + shift` key is pressed on the **first** focusable element (the last element should be focused)

## Why?

I learned that the constrained tabbing [useConstrainedTabbing](https://github.com/WordPress/gutenberg/blob/04802c3e1edef60a6df877d1a5006486564f65da/packages/compose/src/hooks/use-constrained-tabbing/index.js#L33-L76) hook.

For example, when the tab key is pressed on the last focused element, the following logic is used to move focus to the first element:

- Searches for the next element to focus from all tabbable elements
- Add a trap element to the wrapper element if the element is not contained in a wrapper element
- Focus on trap elements
- Browser default behavior moves focus to the next element after the trap element (i.e., the first element in the wrapper)
- Delete the trap element after the focus has successfully moved

However, in Firefox, I found that the trap element is sometimes deleted while the focus remains on the trap element.

```javascript
timeoutId = setTimeout( () => {
	// Sometimes the TRAP element still has focus.
	console.log( document.activeElement );
	node.removeChild( trap );
} );
```

In other words, I believe that the focus is lost by deleting the trap element that remains in focus, and the browser is moving focus without considering the wrapper element.

## How?
I haven't been able to find valid documentation, but I think we need to wait for the browser to properly move focus to the next element before removing the trap element.

So I added a small delay to the `setTimeout` function that removes the trap element.

## Testing Instructions

- Open a modal window such as "Welcome Guide" or "Preferences".
- Confirm that the focus isn't lost from within the modal when you press 'tab' or 'tab+shift' key.
- Or, hold down the tab key to move the focus fast.

## Screenshot

https://user-images.githubusercontent.com/54422211/180595778-5b89e7de-c376-4cae-bbbc-2cef3f4f1377.mp4
